From 47d10f136fb74354f874eab320b78aba384a019e Mon Sep 17 00:00:00 2001
From: Ruslan Bay <67730802+ruslanbay@users.noreply.github.com>
Date: Tue, 17 Feb 2026 05:59:27 +0000
Subject: [PATCH 45/56] media: ipu-isys: Guard cleanup against uninitialized
 resources and add debug logging

Fix NULL pointer dereferences in cleanup functions during driver init failures. Add guards to prevent cleanup of partially-initialized resources:
- Check queue initialization state (vbq.ops)
- Check subdev registration state (v4l2_dev)
---
 drivers/media/pci/intel/ipu-isys-csi2-be-soc.c | 3 +++
 drivers/media/pci/intel/ipu-isys-csi2-be.c     | 3 +++
 drivers/media/pci/intel/ipu-isys-queue.c       | 3 +++
 drivers/media/pci/intel/ipu-isys-tpg.c         | 3 +++
 drivers/media/pci/intel/ipu-isys-video.c       | 3 +++
 drivers/media/pci/intel/ipu4/ipu4-isys-isa.c   | 3 +++
 6 files changed, 18 insertions(+)

diff --git a/drivers/media/pci/intel/ipu-isys-csi2-be-soc.c b/drivers/media/pci/intel/ipu-isys-csi2-be-soc.c
index bf36ca062669..d646597d273d 100644
--- a/drivers/media/pci/intel/ipu-isys-csi2-be-soc.c
+++ b/drivers/media/pci/intel/ipu-isys-csi2-be-soc.c
@@ -222,6 +222,9 @@ void ipu_isys_csi2_be_soc_cleanup(struct ipu_isys_csi2_be_soc *csi2_be_soc)
 {
 	int i;
 
+	if (!csi2_be_soc->asd.sd.v4l2_dev)
+		return;
+
 	v4l2_device_unregister_subdev(&csi2_be_soc->asd.sd);
 	ipu_isys_subdev_cleanup(&csi2_be_soc->asd);
 	for (i = 0; i < NR_OF_CSI2_BE_SOC_STREAMS; i++)
diff --git a/drivers/media/pci/intel/ipu-isys-csi2-be.c b/drivers/media/pci/intel/ipu-isys-csi2-be.c
index bbf1b3ff5f9f..2c1fc7123ad5 100644
--- a/drivers/media/pci/intel/ipu-isys-csi2-be.c
+++ b/drivers/media/pci/intel/ipu-isys-csi2-be.c
@@ -206,6 +206,9 @@ static void csi2_be_set_ffmt(struct v4l2_subdev *sd,
 
 void ipu_isys_csi2_be_cleanup(struct ipu_isys_csi2_be *csi2_be)
 {
+	if (!csi2_be->asd.sd.v4l2_dev)
+		return;
+
 	v4l2_device_unregister_subdev(&csi2_be->asd.sd);
 	ipu_isys_subdev_cleanup(&csi2_be->asd);
 	ipu_isys_video_cleanup(&csi2_be->av);
diff --git a/drivers/media/pci/intel/ipu-isys-queue.c b/drivers/media/pci/intel/ipu-isys-queue.c
index 075cb3ad7e5d..16f9aa97bd0c 100644
--- a/drivers/media/pci/intel/ipu-isys-queue.c
+++ b/drivers/media/pci/intel/ipu-isys-queue.c
@@ -1423,5 +1423,8 @@ int ipu_isys_queue_init(struct ipu_isys_queue *aq)
 
 void ipu_isys_queue_cleanup(struct ipu_isys_queue *aq)
 {
+	if (!aq->vbq.ops)
+		return;
+
 	vb2_queue_release(&aq->vbq);
 }
diff --git a/drivers/media/pci/intel/ipu-isys-tpg.c b/drivers/media/pci/intel/ipu-isys-tpg.c
index 59151f3954ab..10aeec19d09c 100644
--- a/drivers/media/pci/intel/ipu-isys-tpg.c
+++ b/drivers/media/pci/intel/ipu-isys-tpg.c
@@ -251,6 +251,9 @@ static struct media_entity_operations tpg_entity_ops = {
 
 void ipu_isys_tpg_cleanup(struct ipu_isys_tpg *tpg)
 {
+	if (!tpg->asd.sd.v4l2_dev)
+		return;
+
 	v4l2_device_unregister_subdev(&tpg->asd.sd);
 	ipu_isys_subdev_cleanup(&tpg->asd);
 	ipu_isys_video_cleanup(&tpg->av);
diff --git a/drivers/media/pci/intel/ipu-isys-video.c b/drivers/media/pci/intel/ipu-isys-video.c
index 2e09ae9bfd6a..8d53146f116f 100644
--- a/drivers/media/pci/intel/ipu-isys-video.c
+++ b/drivers/media/pci/intel/ipu-isys-video.c
@@ -2317,6 +2317,9 @@ int ipu_isys_video_init(struct ipu_isys_video *av,
 
 void ipu_isys_video_cleanup(struct ipu_isys_video *av)
 {
+	if (!av->aq.vbq.ops)
+		return;
+
 	video_unregister_device(&av->vdev);
 	media_entity_cleanup(&av->vdev.entity);
 	mutex_destroy(&av->mutex);
diff --git a/drivers/media/pci/intel/ipu4/ipu4-isys-isa.c b/drivers/media/pci/intel/ipu4/ipu4-isys-isa.c
index 4ac6545ecc73..a9652b66c1ae 100644
--- a/drivers/media/pci/intel/ipu4/ipu4-isys-isa.c
+++ b/drivers/media/pci/intel/ipu4/ipu4-isys-isa.c
@@ -231,6 +231,9 @@ static struct media_entity_operations isa_entity_ops = {
 
 void ipu_isys_isa_cleanup(struct ipu_isys_isa *isa)
 {
+	if (!isa->asd.sd.v4l2_dev)
+		return;
+
 	v4l2_device_unregister_subdev(&isa->asd.sd);
 	ipu_isys_subdev_cleanup(&isa->asd);
 	ipu_isys_video_cleanup(&isa->av_scaled);
-- 
2.51.0

